File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();

        std::vector<int> res (n,1), 
                        pre (n,1), post (n,1);

        //pre
        for (int i = 1; i < n; i++) {
            pre[i] = pre[i - 1] * nums[i - 1];
        }

        //post
        for (int i = n - 2; i >= 0; i--) {
            post[i] = post[i + 1] * nums[i + 1];
        }

        for (int i = 0; i < n; i++) {
            res[i] = post[i] * pre[i];
        }

        return res;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut right_left: Vec<i32> = vec![1; nums.len()];
        let mut prev_n: i32 = 1;

        for i in (0..nums.len()-1).rev() {
            right_left[i] = nums[i+1] * right_left[i+1];
        }

        for i in 0..nums.len() {
            right_left[i] = right_left[i] * prev_n;
            prev_n *= nums[i];    
        }

        return right_left;
    }
}